compiler/lib/global_deadcode.ml

```diff
@@ -0,0 +1,445 @@
+        | Top -> ()
+        | Values { known; _ } ->
+            Var.Set.iter (* For each known closure value of f *)
+              (fun k ->
+                (* 1. Look at return values, and add edge between x and these values. *)
+                (* 2. Add an edge pairwise between the parameters and arguments *)
+                match global_info.info_defs.(Var.idx k) with
+                | Expr (Closure (params, _)) ->
+                    (* If the function is under/over-applied then global flow will mark arguments and return value as escaping.
+                       So we only need to consider the case when there is an exact application. *)
+                    if List.length params = List.length args
+                    then (
+                      let return_values = Var.Map.find k global_info.info_return_vals in
+                      Var.Set.iter (add_use Propagate x) return_values;
+                      List.iter2 ~f:(add_use Propagate) params args)
+                | _ -> ())
+              known);
+        add_use Compute x f;
+        List.iter
+          ~f:(fun a -> if variable_may_escape a global_info then add_use Compute x a)
+          args
+    | Block (_, vars, _) -> Array.iter ~f:(add_use Compute x) vars
+    | Field (z, _) -> add_use Compute x z
+    | Constant _ -> ()
+    | Closure (_, cont) -> add_cont_deps cont
+    | Prim (_, args) ->
+        List.iter
+          ~f:(fun arg ->
+            match arg with
+            | Pv v -> add_use Compute x v
+            | Pc _ -> ())
+          args
+  in
+  Addr.Map.iter
+    (fun _ block ->
+      (* Add deps from block body *)
+      List.iter
+        ~f:(fun (i, _) ->
+          match i with
+          | Let (x, e) -> add_expr_uses x e
+          | Assign (x, y) -> add_use Compute x y
+          | Set_field (_, _, _) | Offset_ref (_, _) | Array_set (_, _, _) -> ())
+        block.body;
+      (* Add deps from block branch *)
+      match fst block.branch with
+      | Return _ | Raise _ | Stop -> ()
+      | Branch cont -> add_cont_deps cont
+      | Cond (_, cont1, cont2) ->
+          add_cont_deps cont1;
+          add_cont_deps cont2
+      | Switch (_, a) -> Array.iter ~f:add_cont_deps a
+      | Pushtrap (cont, _, cont_h, _) ->
+          add_cont_deps cont;
+          add_cont_deps cont_h
+      | Poptrap cont -> add_cont_deps cont)
+    prog.blocks;
+  uses
+
+(** Return the set of variables used in a given expression *)
+let expr_vars e =
+  let vars = Var.Set.empty in
+  match e with
+  | Apply { f; args; _ } ->
+      let vars = Var.Set.add f vars in
+      List.fold_left ~f:(fun acc x -> Var.Set.add x acc) ~init:vars args
+  | Block (_, params, _) ->
+      Array.fold_left ~f:(fun acc x -> Var.Set.add x acc) ~init:vars params
+  | Field (z, _) -> Var.Set.add z vars
+  | Constant _ -> vars
+  | Closure (params, _) ->
+      List.fold_left ~f:(fun acc x -> Var.Set.add x acc) ~init:vars params
```

Comment:
```suggestion
      vars
```
We want the set of variables which are defined before the expression and used in the expression. This does not include the closure parameters.
The closure may use some variables defined outside, but this will be handled when visiting its body.

Reply:
Make sense, thanks. [238b5de](https://github.com/ocsigen/js_of_ocaml/pull/1503/commits/238b5dee9c01120a4d1a55e821efe6d57fda51b6)